### PR TITLE
JIT: fix remorph assert in cast long shift optimization

### PR DESCRIPTION
**Pull request by AndyAyersMS**
In some cases reachable from IL we may simplify the shift amount,
which sets the MORPHED flag, and then later remorph, leading to an assert
in DEBUG/CHECK builds.

Fix is to clear the MORPHED flag.

Added test case.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@sandreenko PTAL
cc @dotnet/jit-contrib

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -490,6 +490,11 @@ GenTreePtr Compiler::fgMorphCast(GenTreePtr tree)
                 shiftAmount      = gtFoldExpr(shiftAmount);
                 oper->gtOp.gtOp2 = shiftAmount;
 
+#if DEBUG
```
Comment (sandreenko) [inline src/jit/morph.cpp]: I do not think that this workaround is a right fix.

Lets imagine that `shiftAmount` is such expression: "(a>0)? b:b", where `b` is not leaf node. `gtFoldExpr` will return `b`, it is flag will be cleared after your change, but not its children flags. So we will hit the assert again.

I think we should clean `gtFoldExpr` not to call `fgMorphDone` (It doesn't require many changes) or revert the optimization, if it doesn't worth it.

Comment (AndyAyersMS) [reply src/jit/morph.cpp]: That would be true if `gtFoldExpr` did a full top-down bottom-up traversal, but it doesn't do that. So from what I can see in cases where `gtFoldExpr` does walk down one level, `b` is only flagged if it is the ultimate result of folding and if it is a constant.

But I agree with you that not calling `fgMorphTreeDone` at all during `gtFoldxxx` is a better option. Let me look into that and see if it really is a straightforward change.

Comment (sandreenko) [reply src/jit/morph.cpp]: > That would be true if gtFoldExpr did a full top-down bottom-up traversal, but it doesn't do that. So from what I can see in cases where gtFoldExpr does walk down one level, b is only flagged if it is the ultimate result of folding and if it is a constant.

You are right, than the change should work.

Comment (AndyAyersMS) [reply src/jit/morph.cpp]: If you agree, how about we take this fix and the separately work on getting `gtFoldExpr` out of the business of setting the morphed flag?

**Comment by AndyAyersMS** [comment]
Believe these arm failures are a consequence of the issue fixed in #15387. Will retry.

@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test 
@dotnet-bot test Windows_NT armlb Cross Checked Innerloop Build and Test

File: src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [reply src/jit/morph.cpp]: It sound good for me.

**Comment by AndyAyersMS** [comment]
Added #15409 for the flag cleanup.

**Pull request merged**